### PR TITLE
RFC reduce promotion: default to system default

### PR DESCRIPTION
**Pull request by felixrehren**
Before this PR, the [result type of a `map_reduce` is erratic](https://github.com/JuliaLang/julia/issues/20560) on integers.

After this PR, `sum(xs)` for `eltype(xs) = IntX` returns `IntY`, where `Y` the bit-size of the system, when `X <= Y`. I.e. on a 32-bit system, `sum(Int8[])::Int32` and on a 64-bit system `sum(Int8[])::Int64`, etc. Types wider than the system size are untouched, as before. The same changes hold for unsigned integers and `prod`.

This paves the way for `sum(T, xs)::T`, or `sum(zero(T), xs)::T` to be type-stable, see https://github.com/JuliaLang/julia/pull/20561

- [X] adapt tests
- [X] change docs
- [ ] add news
- [X] sums over `Float16` are no longer widened, cf. https://github.com/JuliaLang/julia/issues/5942#issuecomment-233034260
- [ ] ? currently, `widen(::Type{Int8}) = Int32`. Can we change this to `widen(::Type{Int8}) = Int16`? (in line with every other case, where `widen(::Type{IntX}) = Int2X`)

### CODE REVIEW

File: base/reduce.jl [hunk 1]
```diff
@@ -13,20 +13,23 @@ const SmallUnsigned = Union{UInt8,UInt16,UInt32}
 end
 
 const CommonReduceResult = Union{UInt64,UInt128,Int64,Int128,Float32,Float64}
-const WidenReduceResult = Union{SmallSigned, SmallUnsigned, Float16}
+const SmallReduceResult = Union{SmallSigned, SmallUnsigned, Float16}
```
Comment (stevengj) [inline base/reduce.jl]: `SmallReduceResult` seems like a misnomer here.  The result is `Int`, which is *not* small.

`Float16` should still be widened to `Float32`, presumably, which is not a "system size".

Comment (KristofferC) [inline base/reduce.jl]: If `Float16` is a "real computational type"  https://github.com/JuliaLang/julia/issues/5942#issuecomment-233034260 it shouldn't be widened, right?

**Comment by KristofferC** [comment]
- [ ] Run doctests

**Comment by stevengj** [comment]
Bump.

File: base/reduce.jl [hunk 2]
```diff
@@ -12,15 +12,18 @@ const SmallSigned = Union{Int8,Int16,Int32}
...
+promote_sys_size{T}(::Type{T}) = T
+promote_sys_size{T<:SmallSigned}(::Type{T}) = Int
+promote_sys_size{T<:SmallUnsigned}(::Type{T}) = UInt
 # r_promote_type: promote T to the type of reduce(op, ::Array{T})
 # (some "extra" methods are required here to avoid ambiguity warnings)
 r_promote_type{T}(op, ::Type{T}) = T
-r_promote_type{T<:WidenReduceResult}(op, ::Type{T}) = widen(T)
-r_promote_type{T<:WidenReduceResult}(::typeof(+), ::Type{T}) = widen(T)
-r_promote_type{T<:WidenReduceResult}(::typeof(*), ::Type{T}) = widen(T)
+r_promote_type{T<:WidenReduceResult}(op, ::Type{T}) = promote_sys_size(T)
```
Comment (stevengj) [inline base/reduce.jl]: As noted in #21523, widening is not necessarily appropriate for an arbitrary user-defined operator.   Maybe the default should be `T`.

**Comment by felixrehren** [comment]
Will try and do this in June (changed jobs recently, hence different availability to before). If I understand the consensus correctly, it's to remove any promotion from `reduce`, but keep it for `sum`. Confirmation (e.g. :+1:/:-1:) appreciated on this.

With apologies, the thing that held this up in the first place is that I couldn't figure out how to run doctests. @KristofferC

**Comment by TotalVerb** [comment]
@felixrehren, are you still able to work on this?

cc @StefanKarpinski

**Comment by KristofferC** [comment]
Don't let the doctests hold you up then, they can be fixed later.

**Comment by TotalVerb** [comment]
Hey @felixrehren, I've continued this PR in #22825. Thanks for your work on it!

**Pull request closed without merging**